### ISSUE TITLE
Segfault while trying to join a MUC

### ISSUE THREAD

**Issue by elimohl** [gkdr/lurch#65]
It only happens with one certain MUC and one certain pidgin profile (doesn't happen when I run pidgin with another home directory) but it happens every time I try to join.
First time it has happened when I've typed `/lurch enable` to this MUC.

Run on
Debian Sid
Pidgin 2.12.0 (libpurple 2.12.0)

```
(gdb) run
Starting program: /usr/bin/pidgin
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffdbb98700 (LWP 23045)]
[New Thread 0x7fffd8a27700 (LWP 23053)]
[New Thread 0x7fffd3fff700 (LWP 23054)]

Thread 1 "pidgin" received signal SIGSEGV, Segmentation fault.
strnlen () at ../sysdeps/x86_64/strlen.S:103
103 ../sysdeps/x86_64/strlen.S: No such file or directory.
(gdb) bt
#0  0x00007ffff4958830 in strnlen () at ../sysdeps/x86_64/strlen.S:103
#1  0x00007fffe2546314 in lurch_message_decrypt (gc_p=0x100c95170, msg_stanza_pp=0x7fffffffc090) at ./src/lurch.c:1934
#2  0x00007fffe2546ad5 in lurch_xml_received_cb (gc_p=0x100c95170, stanza_pp=0x7fffffffc090) at ./src/lurch.c:2097
#3  0x00007ffff52268f0 in purple_signal_emit_vargs (instance=<optimized out>, signal=signal@entry=0x7fffe78ff076 "jabber-receiving-xmlnode", args=args@entry=0x7fffffffbf70)
    at ././libpurple/signals.c:482
    #4  0x00007ffff5226a4e in purple_signal_emit (instance=<optimized out>, signal=signal@entry=0x7fffe78ff076 "jabber-receiving-xmlnode") at ././libpurple/signals.c:434
    #5  0x00007fffe78e2b46 in jabber_process_packet (js=js@entry=0x100c956a0, packet=packet@entry=0x7fffffffc090) at ././libpurple/protocols/jabber/jabber.c:335
    #6  0x00007fffe78ef8d4 in jabber_parser_element_end_libxml (user_data=0x100c956a0, element_name=<optimized out>, prefix=<optimized out>, namespace=<optimized out>)
        at ././libpurple/protocols/jabber/parser.c:169
        #7  0x00007fffeffc9043 in  () at /usr/lib/x86_64-linux-gnu/libxml2.so.2
        #8  0x00007fffeffd1907 in  () at /usr/lib/x86_64-linux-gnu/libxml2.so.2
        #9  0x00007fffeffd336b in xmlParseChunk () at /usr/lib/x86_64-linux-gnu/libxml2.so.2
        #10 0x00007fffe78efd7d in jabber_parser_process (js=0x100c956a0, buf=<optimized out>, len=<optimized out>) at ././libpurple/protocols/jabber/parser.c:279
        #11 0x00007fffe78deeda in jabber_recv_cb_ssl (data=0x100c95170, gsc=0x100d18e50, cond=<optimized out>) at ././libpurple/protocols/jabber/jabber.c:661
        #12 0x0000000100071bde in pidgin_io_invoke (source=<optimized out>, condition=<optimized out>, data=0x100c953e0) at ././pidgin/gtkeventloop.c:73
        #13 0x00007ffff551eaaa in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
        #14 0x00007ffff551ee60 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
        #15 0x00007ffff551f182 in g_main_loop_run () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
        #16 0x00007ffff67c93b7 in gtk_main () at /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
        #17 0x0000000100038322 in main (argc=<optimized out>, argv=<optimized out>) at ././pidgin/gtkmain.c:939
```

**Comment by gkdr** [comment]
I think `strnlen()` segfaults on nullpointers, and the way I see this happening is in the following line: https://github.com/gkdr/lurch/blob/83048a5e4ef5ddff1b5dce4509356f87ba76dd67/src/lurch.c#L1896

According to the manpage, `strdup()` returns null when there is no sufficient memory, but I would guess that it also returns null when the initial string is null. In this case this means that the user's presence information was not parsed and the JID is not available.
Why do you receive a message on MUC join? Is there some sort of bot? Does the MUC have a lot of users? Is it anonymous?

**Comment by elimohl** [comment]
>I would guess that it also returns null when the initial string is null

[It seems it's true.](https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html#g-strdup)

The MUC doesn't have a lot of users and there is no bot, but it is anonymous. When I join the MUC with lurch disabled, I recieve `<presence>` and then last messages from the MUC.

**Comment by gkdr** [comment]
If it is anonymous, then the lookup will of course fail, because the JID could not be parsed from the presence. But now I wonder why it works for you if you use a different profile, and if I really didn't think about checking for something like that. I can't quite remember if it's possible to easily see if a MUC is anonymous or not, but it definitely shouldn't segfault. What happens when you use the other profile which seems to work?

**Comment by elimohl** [comment]
I cannot reproduce the segfault on join anymore.
Now I get `There was an error encrypting the message and it was not sent. You can try again, or try to find the problem by looking at the debug log.` when I try to send a message whether a conference is private or not.

From Debug Window:
```
(00:58:51) lurch: lurch_message_encrypt_groupchat: no entry in ht for 
<conference name>
 (0)
(00:58:51) g_log: xmlnode_to_str_helper: assertion 'node != NULL' failed
(00:58:51) g_log: jabber_send_raw: assertion 'data != NULL' failed
```

But I get a segfault with another scenario. If conference is not anonymous on the start but an owner change settings to make it anonymous and then send a message, my pidgin segfaults on recieving this message.

**Comment by elimohl** [comment]
Just got a segfault while try to join to MUC for which I don't remember enable lurch at all. And become able to join only after I've disabled lurch in plugings setting.

**Comment by gkdr** [comment]
I'm sorry this doesn't work well, but I'm really not sure why you get all those weird hashtable lookup errors. Could it be an issue with a non-latin script?

**Comment by elimohl** [comment]
It is all the same when there is only latin symbols in a chat.
Joining fails only for specific MUCs (and now it stops failing again), Encrypting messages fails for every MUC.

**Comment by gkdr** [comment]
Can you tell me if the contents of the "cl" table in your `..._omemo_db.sqlite` file looks weird? Does it fail with the same message on the debug log as the one you showed me already?

**Comment by salexan2001** [comment]
I just had the same problem on my machine, so I thought it might be useful to add this information here.

My system is an up to date Arch Linux with Pidgin 2.12.0 and I tried both, libpurple-lurch-git and libpurple-lurch from the AUR.

The following trace was extracted when trying to join a group chat with three members. The segfault occurred just after double clicking the chat room.

```
(gdb) run
Starting program: /usr/bin/pidgin 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7fffd86bb700 (LWP 13653)]
[New Thread 0x7fffd4ea4700 (LWP 13654)]
[New Thread 0x7fffcffff700 (LWP 13655)]
[New Thread 0x7fffcf7fe700 (LWP 13656)]
[Thread 0x7fffcf7fe700 (LWP 13656) exited]

Thread 1 "pidgin" received signal SIGSEGV, Segmentation fault.
0x00007ffff48f36c0 in __strnlen_sse2 () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff48f36c0 in __strnlen_sse2 () at /usr/lib/libc.so.6
#1  0x00007fffd5bb1246 in  () at /usr/lib/purple-2/lurch.so
#2  0x00007ffff520fc8a in purple_signal_emit_vargs () at /usr/lib/libpurple.so.0
#3  0x00007ffff520fdff in purple_signal_emit () at /usr/lib/libpurple.so.0
#4  0x00007fffd86ef698 in jabber_process_packet () at /usr/lib/purple-2/libjabber.so.0
#5  0x00007fffd86fca25 in  () at /usr/lib/purple-2/libjabber.so.0
#6  0x00007fffefd3363d in  () at /usr/lib/libxml2.so.2
#7  0x00007fffefd39b09 in  () at /usr/lib/libxml2.so.2
#8  0x00007fffefd3b43e in xmlParseChunk () at /usr/lib/libxml2.so.2
#9  0x00007fffd86fcf2e in jabber_parser_process () at /usr/lib/purple-2/libjabber.so.0
#10 0x00007fffd86eb90d in  () at /usr/lib/purple-2/libjabber.so.0
#11 0x00005555555c564f in  ()
#12 0x00007ffff552a0be in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#13 0x00007ffff552bf69 in  () at /usr/lib/libglib-2.0.so.0
#14 0x00007ffff552cf42 in g_main_loop_run () at /usr/lib/libglib-2.0.so.0
#15 0x00007ffff67ce3a7 in gtk_main () at /usr/lib/libgtk-x11-2.0.so.0
#16 0x00005555555892ad in main ()
```

I also had a look at the cl table in "..._omemo_db.sqlite". It does not look weird, it just does not contain any rows. Is there anything else I can contribute for solving the problem?

Maybe:
pidgin is running here on the xfce4-desktop (4.12) and I use gtk 3.22.26 (pidgin seems to use gtk 2.24.31).
glibc version is 2.26

**Comment by gkdr** [comment]
Can you tell me if the chat room is set to anonymous?

**Comment by salexan2001** [comment]
I don't think it's anonymous. But I am not completely sure what an anonymous chat room is and how to check. (I tried to find here: https://xmpp.org/extensions/xep-0045.html) Is it possible to get this info in pidgin?

**Comment by gkdr** [comment]
"Anonymous" means that chat participants cannot see each other's JID, just the nick (even if everyone is in everyone's buddy list). The JID is necessary to fetch the necessary key data. Since Pidgin doesn't provide plugins with much information about chat participants, I parse the  `presence` stanzas myself and I suspect I didn't add enough `null` checks for when the room is actually anonymous and there is no JID information.

In Pidgin, checking this (or any other) MUC property is unfortunately quite hidden, and I'm not sure if you can even do it if you're not a channel moderator. A channel moderator has to open the MUC window (I guess since Pidgin crashes with lurch active you have to deactivate it first if you are a moderator) and then type `/config`.  There it says "Who may discover real JIDs?", and "moderators only" means anonymous.
When trying to find this information just now I realized that I did in fact join an anonymous MUC with lurch active, so I can't reproduce the issue so far.

**Comment by salexan2001** [comment]
Thanks for the explanation!

I just checked and found the following setting "Present real Jabber IDs to:" which is set to "anyone". So this room is not anonymous.

Do you have any other idea why it might crash or how we can proceed to find out?

I am a moderator of this chatroom and created it using the android client conversations.

Could the server software cause it? (It is running ejabberd 14.07 with some important XEPs disabled.)

**Comment by gkdr** [comment]
A stack trace with line numbers like in the OP would be quite helpful. In the `makefile`, the `-g` flag is included, so I'm not sure why you didn't get them, but you could try compiling yourself if you just installed from the AUR before.

**Comment by salexan2001** [comment]
I just compiled directly from this git-repository and tried to reproduce the problem and it seems to be gone (no segfault anymore) which is weird, because the sourcecode seems to be the same.

However it still fails to send encrypted messages causing the debug-output:
```
(13:01:40) lurch: lurch_message_encrypt_groupchat: no entry in ht for ...@...!
 (0)
(13:01:40) g_log: xmlnode_to_str_helper: assertion 'node != NULL' failed
(13:01:40) g_log: jabber_send_raw: assertion 'data != NULL' failed
```
(Two users in the room. I had activated omemo using '/lurch enable')

But I guess, that's a different problem?

**Comment by NattyNarwhal** [comment]
I believe I can reproduce this on Windows as well, but I don't have any debugging information, unfortunately.

**Comment by NattyNarwhal** [comment]
To add to that, I get the behaviours originally referenced: joining MUC rooms crashes Pidgin. If I disable the plugin, join the MUC, and re-enable the plugin, then `/lurch enable` in the MUC, I get the error @salexan2001 mentions trying to send messages.

**Comment by EionRobb** [comment]
@NattyNarwhal if you could follow https://developer.pidgin.im/wiki/GetABacktrace to get the backtrace from the Pidgin.rpt file, it should help track down whether its the same issue or a different one

**Comment by NattyNarwhal** [comment]
```
Error occured on Wednesday, March 14, 2018 at 22:53:00.

Windows Version 6.2 Build 9200 

C:\Program Files (x86)\Pidgin\pidgin.exe caused an Access Violation at location 75bc8640 in module C:\WINDOWS\System32\msvcrt.dll Reading from location 00000000.

Registers:
eax=00000000 ebx=0061e9f4 ecx=00000000 edx=00000002 esi=0061eb18 edi=0061ea00
eip=75bc8640 esp=0061e9bc ebp=0061ea78 iopl=0         nv up ei pl zr na po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00210246

Call stack:
         C:\WINDOWS\System32\msvcrt.dll [7.0.16299.125]
75BC8640 C:\WINDOWS\System32\msvcrt.dll  strlen
0546F41A C:\Program Files (x86)\Pidgin\plugins\lurch.dll
         C:\Program Files (x86)\Pidgin\libpurple.dll [2.12.0.0]
6C01F907 C:\Program Files (x86)\Pidgin\libpurple.dll  purple_signal_emit_vargs
         C:\WINDOWS\System32\msvcrt.dll [7.0.16299.125]
75B87084 C:\WINDOWS\System32\msvcrt.dll  free

```

And if I redirect stderr, last item is: `(22:53:00) GLib: g_hash_table_lookup: assertion `hash_table != NULL' failed`

**Comment by NattyNarwhal** [comment]
Interestingly, an older crash from when I tried the compat version had richer symbols:

```
Error occured on Wednesday, March 14, 2018 at 20:58:31.

Windows Version 6.2 Build 9200 

C:\Program Files (x86)\Pidgin\pidgin.exe caused an Access Violation at location 75bc8640 in module C:\WINDOWS\System32\msvcrt.dll Reading from location 00000000.

Registers:
eax=00000000 ebx=0061e9f4 ecx=00000000 edx=00000002 esi=0061eb18 edi=0061ea00
eip=75bc8640 esp=0061e9bc ebp=0061ea78 iopl=0         nv up ei pl zr na po nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00210246

Call stack:
         C:\WINDOWS\System32\msvcrt.dll [7.0.16299.125]
75BC8640 C:\WINDOWS\System32\msvcrt.dll  strlen
6EF4F41A C:\Program Files (x86)\Pidgin\plugins\lurch-compat.dll  lurch_xml_received_cb  C:\cygwin\home\Eion\lurch/src/lurch.c:2097
         C:\Program Files (x86)\Pidgin\libpurple.dll [2.12.0.0]
60D1F907 C:\Program Files (x86)\Pidgin\libpurple.dll  purple_signal_emit_vargs
         C:\WINDOWS\System32\msvcrt.dll [7.0.16299.125]
75B87084 C:\WINDOWS\System32\msvcrt.dll  free
```

**Comment by mejo-** [comment]
I seem to discover the same bug when joining a lurch-enabled MUC. After some testing it became obvious that it's related to lurch messages in the channel history. Everytime I joined the channel with lurch enabled, pidgin crashed.

Even disabling OMEMO for the channel (I removed it from the 'cl' table in `~/.purple/account@server_omemo_db.sqlite`), pidgin kept crashing at joining the channel. What finally helped, was clearing the channel history. Since then, pidgin doesn't crash anymore even though lurch is enabled.

**Comment by gkdr** [comment]
I still cannot seem to reproduce this. What I tried:

1. Make sure MUC history is enabled
2. Join the room _test_ with user _b_, then disconnect the account
3. Write to the MUC, both with lurch enabled and disabled, with the following users:
  -- _a_: in user _b_'s buddy list and there exists an OMEMO session
  -- _d_: **not** in user _b_'s buddy list, and therefore no OMEMO session
4. Part the room with both _a_ and _d_
5. Reconnect _b_

As part of the MUC history, I only received the plaintext messages (from the server), which did not crash me even if the user was not in the contact list as plaintext messages are simply ignored. I am running Prosody 0.9.12, and joining the room gives me the following raw XMPP:

```XML
<presence to='b@localhost/d16b6331-430c-40e5-9ed0-255db1e8656f' from='test@conference.localhost/b'><priority>1</priority><c hash='sha-1' ext='voice-v1 camera-v1 video-v1' ver='SeLBmLU1xeoYKJyZZnOf9E9AJuY=' node='http://pidgin.im/' xmlns='http://jabber.org/protocol/caps'/><x xmlns='http://jabber.org/protocol/muc#user'><item jid='b@localhost/d16b6331-430c-40e5-9ed0-255db1e8656f' affiliation='member' role='participant'/><status code='110'/><status code='100'/></x></presence><message id='purplecdf1005e' type='groupchat' to='b@localhost/d16b6331-430c-40e5-9ed0-255db1e8656f' from='test@conference.localhost/a'><body>asdasdas</body><request xmlns='urn:xmpp:receipts'/><delay stamp='2018-04-02T11:05:19Z' xmlns='urn:xmpp:delay'/><x stamp='20180402T11:05:19' xmlns='jabber:x:delay'/></message><message id='purplea6f85775' type='groupchat' to='b@localhost/d16b6331-430c-40e5-9ed0-255db1e8656f' from='test@conference.localhost/b'><body>heyyyyy</body><request xmlns='urn:xmpp:receipts'/><delay stamp='2018-04-02T11:05:25Z' xmlns='urn:xmpp:delay'/><x stamp='20180402T11:05:25' xmlns='jabber:x:delay'/></message><message id='purple2080e233' type='groupchat' to='b@localhost/d16b6331-430c-40e5-9ed0-255db1e8656f' from='test@conference.localhost/d'><body>testing ;))))</body><request xmlns='urn:xmpp:receipts'/><delay stamp='2018-04-02T11:10:10Z' xmlns='urn:xmpp:delay'/><x stamp='20180402T11:10:10' xmlns='jabber:x:delay'/></message><message type='groupchat' to='b@localhost/d16b6331-430c-40e5-9ed0-255db1e8656f' from='test@conference.localhost/a'><subject>testt</subject></message>
```
@mejo-, can you clarify what you mean by "channel history"? If it is just the MUC history feature, can you show me the raw XML you receive on join?

@NattyNarwhal, @salexan2001: I realized that I only had the `-g` flag in the makefile on the dev branch. Now it's in the master too, so if you run 0.6.6 through gdb, you should see better debug output such as the exact lines, if you want to try again.

**Comment by NattyNarwhal** [comment]
I managed to reproduce it by toggling OMEMO on and off for the room, having a mix of OMEMO and non-OMEMO messages in the history. It seems erratic to get a test case because there's so many moving parts going on.

**Comment by mejo-** [comment]
In my case, a mix of OMEMO and non-OMEMO messages in history and different clients (Gajim 1.0, Pidgin 2.12.0 and Pidgin 2.13.0) were involved. Both Pidgin 2.12.0 and 2.13.0 crashed the same way.

**Comment by gkdr** [comment]
I suspect it could be one or both of the following:
- Other MUC implementations (e.g. for ejabberd) also save `<message>`s without a `<body>`
- Other clients (like Gajim) additionally include a `<body>` tag

If you can consistently reproduce it with a room, I don't need the exact setup! I just need that first `<presence>` stanza you receive when you join, it should look about the same as what I posted.

**Comment by mejo-** [comment]
@gkdr: And yes, with channel history I mean the MUC history feature. Maybe also MUC-MAM, but not sure here. Unfortunately I'm rather busy right now, so I cannot give any estimation on when I will find time to reproduce it. We disabled OMEMO for MUCs for now as it broke our communication.

**Comment by NattyNarwhal** [comment]
The room I was using used Prosody. The channel had Conversations and Gajim usage, at least.

If I open up the XMPP console extension, the hang happens right after sending the initial presence burst for join. Let me try to disable Lurch and see what message is the first received.

**Comment by NattyNarwhal** [comment]
Ok, the first presence stanza received after the burst for MUC:

```xml
<presence to='calvin@snacki.ca/azimuth' from='stratcom@group.snacki.ca/calvin'>
	<status>XMPP: It's not a meme anymore, honest</status>
	<priority>1</priority>
	<c xmlns='http://jabber.org/protocol/caps' ver='fEjIiGqjFlDgCVnPu+oCQJdkvhw=' hash='sha-1' node='http://pidgin.im/'/>
	<x xmlns='http://jabber.org/protocol/muc#user'>
		<item jid='calvin@snacki.ca/azimuth' affiliation='owner' role='moderator'/>
		<status code='110'/>
		<status code='100'/>
	</x>
</presence>
```

**Comment by gkdr** [comment]
@mejo- Since Pidgin doesn't know MAM I suspect it should simply be ignored, but I guess I'll look into it too.

@NattyNarwhal Thanks for your help, and sorry, but I didn't pay attention. In the debug log (not XMPP console) it's just one big burst so I thought it's the same. Right after the first presence, the MUC actually sends regular `<message>` stanzas. These I need to see.

**Comment by NattyNarwhal** [comment]
Ok, the first message in said room:

```xml
<message id='4532a092-31a1-4810-8e2d-ee59e032da67' type='groupchat' to='calvin@snacki.ca/azimuth' from='stratcom@group.snacki.ca/calvin'>
	<thread>None</thread>
	<active xmlns='http://jabber.org/protocol/chatstates'/>
	<origin-id xmlns='urn:xmpp:sid:0' id='4532a092-31a1-4810-8e2d-ee59e032da67'/>
	<encrypted xmlns='eu.siacs.conversations.axolotl'>
		<header sid='2111591022'>
			<key rid='1916986802'>MwohBdHdcM6XWyMFU3K6DQdxBYalNCF//Ox6+LlHEM0bncoKEAoYAyIw0NXATCczmDjpPBUO5CXV/vP8N0kJ1zorL7URlCuDf5gXwp5PI+KXVCVBax4IMWp+0eoylXSlVc4=</key>
			<key rid='1362820266'>MwohBY8pdyBv118xrkwjdNxY2fNun3536pX0l2Fg/FutiOM6EAEYACIwfduxn4ZqhUeBHtD4uwcgyGkpRkgR/WFp4C0Ye+BQUI6W/swfUfpylAerhg6ht+0YJtCD6zoy768=</key>
			<key prekey='true' rid='1529591603'>MwhXEiEFu/QWxNQPMZA7DgVyvppWHMbfFAdyn/QDnzz92xs500QaIQWLNj7Dvl5dLkxksjgBsDXoHy6LKOi6CN4+k7Iv+XbHDiJiMwohBeuX2AFtWeBgegyCXYjzpYX3Q/TjEqm7RP+9MajEdp0wECIYACIwRVaSYPgzPIDldDxPbSouXjH+Xm4Cz1a3+boqT4oyYvpJll3/vAL2qKxIejItoQfPaZ67fuBTGVco66Tx7g8wAQ==</key>
			<key rid='470196580'>MwohBTerCahVU+Ukc9Nq8r1zbaGZ3J08r03PiIRrnPGHia8zEA8YFyIwQHoQKzHOD0VNaGXL64ca/Fh/rlyXimjvd6+1GsIEGkCLdlzoiKXNpP2+IN4vhfWHd9fcP5Gdm8Q=</key>
			<key rid='541634781'>MwohBXgBDginW5bf263MfmOUDpTkdXkrYdb9yUtvosJwLfVgEDEYACIwKODdyvUlvp0SZMVslozxuCYJaVUrT1hasdG0fLeNXit62dKuENP6k/42pol3hy6ZzKPsuRdJvcw=</key>
			<key prekey='true' rid='687547638'>MwgzEiEFQHRZ0MwwpmGF/gbDvHCul8bMZqlXvjw136W5lVTAtWIaIQWLNj7Dvl5dLkxksjgBsDXoHy6LKOi6CN4+k7Iv+XbHDiJiMwohBb9Nh+NIh/temrlBESs1JMNqyJfocC16B7vQk1RdqVwBEB0YACIw3Lsi4EnXQi46dHnsuRJpzB//C0dR+p/Kf6g9pIpdyxXGW2JIKXuyquRL5N+3N+G+Foiy++5l/kAo66Tx7g8w+8qmgAQ=</key>
			<iv>OhXvyeR+z2QRRC2G95TEeA==</iv>
		</header>
		<payload>wV9JzsdZFej4LTB9/7roGkjOE3Ma</payload>
	</encrypted>
	<encryption xmlns='urn:xmpp:eme:0' namespace='eu.siacs.conversations.axolotl' name='OMEMO'/>
	<body>You received a message encrypted with OMEMO but your client doesnt support OMEMO.</body>
	<store xmlns='urn:xmpp:hints'/>
	<x xmlns='http://jabber.org/protocol/muc#user'>
		<item jid='calvin@snacki.ca' affiliation='owner' role='moderator'/>
	</x>
	<delay xmlns='urn:xmpp:delay' stamp='2018-03-16T01:25:28Z' from='stratcom@group.snacki.ca'/>
	<stanza-id xmlns='urn:xmpp:sid:0' id='4VcEr86Po4GM-sCn' by='stratcom@group.snacki.ca'/>
</message>
```

A different one:

```xml
<message id='377b2b10-ec6f-40fd-bc1e-72de934e9817' type='groupchat' to='calvin@snacki.ca/azimuth' from='stratcom@group.snacki.ca/calvin'>
	<body>I sent you an OMEMO encrypted message but your client doesn’t seem to support that. Find more information on https://conversations.im/omemo</body>
	<encrypted xmlns='eu.siacs.conversations.axolotl'>
		<header sid='470196580'>
			<key rid='541634781'>MwohBZGCBsQXLp85roi6RdXcFEStSJe/O5CwixdPalXFPaQkEAYYACIwoScXjCnBDufUpHIGk6OFp6p+uSE/m1e0f1966dO4tfS8uh2/eCyk0nt2lJ6HkI8x0Stk+un5KVg=</key>
			<key prekey='true' rid='687547638'>MwgjEiEFf/OJPXenXFy9KLjyS6HgeAWFV43FIUj/dLgUZClAVgUaIQXNiXnBa++Nc2oDyrF4kHJaWylju2FBsPtI1+8ANVzWQyJiMwohBYZmLuTSccZXXbPKKceXmfuqbgpzngZbn21pobjROQM5EAYYACIw6Ugh27tu1dQBTsWH9UnwRXnW/oXrDCxXlnpLq7BmABCaCc4W3B9YcopqCkkZyznJhVS9oPnQsmAo5MKa4AEw+8qmgAQ=</key>
			<key rid='1362820266'>MwohBWQXI10S96gqb7x+QdhzflTsvuBERJ9tBQRurQB0Vll4EAAYAiIwOdtnlnd+OpqjyGIF99w6ADgB5+T17LyHOIaZ4Hv+KzI6upx7Ce+c0iBQkV/VpYGqCx0Ox33NUSs=</key>
			<key rid='1529591603'>MwohBdB/NkkjSgvSVPHZZyYe863cqMOiwYb23xAy3/OLFf0LEAYYAyIw1nv7CdYAjK+KQPTSdgnQD2RRhPd+6Fb+ccjnrc18dvsB6PNR/hTsgpKQbwmjmmXaEOp1aEcZ7jA=</key>
			<key rid='1916986802'>MwohBfB/Tn1dlD84kQqIWuoql2WCF37mRlY1JeWkstns9ZR/EAMYASIwmPFrUBGGBszp9E/LXbWscdeWDn3NeBaqi3MrP1CriScU//FDfrMfEjuKtuIMsuCa68y2pgn2Pp0=</key>
			<key rid='2111591022'>MwohBY0X2jfR1i09c7HiV1UROAfdQcOmUI43u2hfUwjsmI82EAAYACIwh5Y9HvL2DYRVj5JUpNGyYldmYeE/lzZPuUjQoeyhegA1ImqLySO4uFqdHhFHlOg1wdewZXl66QE=</key>
			<iv>va+oBonspJcMq5nbL0asvw==</iv>
		</header>
		<payload>qGE/x3Dv+565b8/KvDqS+40fziZErQQ1KWsU</payload>
	</encrypted>
	<markable xmlns='urn:xmpp:chat-markers:0'/>
	<origin-id xmlns='urn:xmpp:sid:0' id='377b2b10-ec6f-40fd-bc1e-72de934e9817'/>
	<store xmlns='urn:xmpp:hints'/>
	<encryption xmlns='urn:xmpp:eme:0' namespace='eu.siacs.conversations.axolotl' name='OMEMO'/>
	<x xmlns='http://jabber.org/protocol/muc#user'>
		<item jid='calvin@snacki.ca' affiliation='owner' role='moderator'/>
	</x>
	<delay xmlns='urn:xmpp:delay' stamp='2018-03-18T13:46:49Z' from='stratcom@group.snacki.ca'/>
	<stanza-id xmlns='urn:xmpp:sid:0' id='tTWLTsQxmNEdBEoy' by='stratcom@group.snacki.ca'/>
</message>
```

**Comment by gkdr** [comment]
Okay, thanks a lot. I do not add a `<body>` because last time I checked, ChatSecure couldn't handle it (see #57), but it seems other clients add it, and even thought the issue is still open I'll ask the other devs about it.
Since there is a `<body>` element, it is saved to history, and then there are in fact OMEMO messages in there which I probably don't handle correctly. I'll try reproducing it by also using Gajim, but now I definitely know what to look out for. Thanks a lot!

**Comment by gkdr** [comment]
Gajim won't let me install the OMEMO plugin with an error, probably because my system and everything on it is really outdated (I have been hoping for some time now that I can just buy a new laptop, but somehow it's not happening... ). Then I remembered that I was trying to be smart and added options to include the `<body>` in _libomemo_, so I just have to change the method call inside _lurch_ with no code changes required. Well, it doesn't work, see here: https://github.com/gkdr/libomemo/issues/20

It seems I'll have to fix that first, before I can even try to reproduce it. But I'm on it.

**Comment by gkdr** [comment]
Managed to reproduce it. Now I can finally start fixing it.

**Comment by mejo-** [comment]
That's great news, @gkdr. Thanks for investigating the issue. I'll gladly test any fix once it's available.

**Comment by gkdr** [comment]
It took me so long because I thought I'd do it "properly", but now I think it's a different issue to be able to really handle the MUC history.
If you can build from the dev branch, please try it and tell me if it works. Instead of displaying the body I could also remove the message completely, which might be less annoying, but which would hide the history until it is properly supported. Tell me what you prefer.

**Comment by mejo-** [comment]
Thanks a ton @gkdr for working on the issue. I will give the dev branch a try shortly and report back.

Regarding displaying the MUC history, I would very much prefer to see the message bodies :)

**Comment by gkdr** [comment]
As I said in several other places, I rewrote the MUC backend, so this shouldn't happen at all. I tried to copy the behaviour but I frankly was too lazy to construct this test case again, so if the messages bodies don't show up please open another issue as I'm (finally!) going to close this one.

**Comment by mejo-** [comment]
I just wanted to confirm that the Pidgin crashes are fixed for me with lurch 0.6.7. Thanks a ton for that.

MUC history still *is* pretty much broken with OMEMO turned on. But I'll open a separate issue on that with further details as soon as I find time.
